keylime-agent/src/payloads.rs

```diff
@@ -0,0 +1,481 @@
+pub(crate) fn write_out_key_and_payload(
+    dec_payload: &[u8],
+    dec_payload_path: &Path,
+    key: &SymmKey,
+    key_path: &Path,
+) -> Result<()> {
+    let mut key_file = fs::File::create(key_path)?;
+    let bytes = key_file.write(key.bytes())?;
+    if bytes != key.bytes().len() {
+        return Err(Error::Other(format!("Error writing symm key to {:?}: key len is {}, but {} bytes were written", key_path, key.bytes().len(), bytes)));
+    }
+    info!("Wrote payload decryption key to {:?}", key_path);
+
+    let mut dec_payload_file = fs::File::create(dec_payload_path)?;
+    let bytes = dec_payload_file.write(dec_payload)?;
+    if bytes != dec_payload.len() {
+        return Err(Error::Other(format!("Error writing decrypted payload to {:?}: payload len is {}, but {} bytes were written", dec_payload_path, dec_payload.len(), bytes)));
+    }
+    info!("Wrote decrypted payload to {:?}", dec_payload_path);
+
+    Ok(())
+}
+
+// run a script (such as the init script, if any) and check the status
+pub(crate) async fn run(dir: &Path, script: &str, uuid: &str) -> Result<()> {
+    let script_path = dir.join(script);
+    info!("Running script: {:?}", script_path);
+
+    if !script_path.exists() {
+        info!("No payload script {} found in {}", script, dir.display());
+        return Ok(());
+    }
+
+    if fs::set_permissions(&script_path, fs::Permissions::from_mode(0o700))
+        .is_err()
+    {
+        return Err(Error::Other(format!(
+            "unable to set {:?} as executable",
+            &script_path
+        )));
+    }
+
+    info!("Executing payload script: {}", script_path.display());
+
+    match Command::new("sh")
+        .arg("-c")
+        .arg(script_path.to_str().unwrap()) //#[allow_ci]
+        .current_dir(dir)
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .status()
+        .await
+    {
+        Ok(_) => {
+            info!("{:?} ran successfully", &script_path);
+            Ok(())
+        }
+        Err(e) => Err(Error::Other(format!(
+            "{:?} failed during run: {}",
+            &script_path, e
+        ))),
+    }
+}
+
+// checks if keylime-agent.conf indicates the payload should be unzipped, and does so if needed.
+// the input string is the directory where the unzipped file(s) should be stored.
+pub(crate) fn optional_unzip_payload(
+    unzipped: &Path,
+    config: &config::KeylimeConfig,
+) -> Result<()> {
+    if config.agent.extract_payload_zip {
+        let zipped_payload = &config.agent.dec_payload_file;
+        let zipped_payload_path = unzipped.join(zipped_payload);
+
+        info!("Unzipping payload {} to {:?}", &zipped_payload, unzipped);
+
+        let mut source = fs::File::open(&zipped_payload_path)?;
+        uncompress_archive(&mut source, unzipped, Ownership::Ignore)?;
+    }
+
+    Ok(())
+}
+
+pub(crate) async fn run_encrypted_payload(
```

Comment:
Does this need to be `pub(crate)`, while it is only used in this file? Maybe good to review other functions as well.

Reply:
I reviewed the files I modified to make public only the functions that should be accessible externally.